extensions-contrib/scan-query/src/main/java/io/druid/query/scan/ScanQueryRunnerFactory.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.druid.query.scan;
+
+import com.google.common.base.Function;
+import com.google.inject.Inject;
+import io.druid.common.utils.JodaUtils;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.guava.Sequence;
+import io.druid.java.util.common.guava.Sequences;
+import io.druid.query.Query;
+import io.druid.query.QueryContextKeys;
+import io.druid.query.QueryRunner;
+import io.druid.query.QueryRunnerFactory;
+import io.druid.query.QueryToolChest;
+import io.druid.segment.Segment;
+
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+
+public class ScanQueryRunnerFactory implements QueryRunnerFactory<ScanResultValue, ScanQuery>
+{
+  private final ScanQueryQueryToolChest toolChest;
+  private final ScanQueryEngine engine;
+
+  @Inject
+  public ScanQueryRunnerFactory(
+      ScanQueryQueryToolChest toolChest,
+      ScanQueryEngine engine
+  )
+  {
+    this.toolChest = toolChest;
+    this.engine = engine;
+  }
+
+  @Override
+  public QueryRunner<ScanResultValue> createRunner(Segment segment)
+  {
+    return new ScanQueryRunner(engine, segment);
+  }
+
+  @Override
+  public QueryRunner<ScanResultValue> mergeRunners(
+      ExecutorService queryExecutor,
+      final Iterable<QueryRunner<ScanResultValue>> queryRunners
+  )
+  {
+    // in single thread and in jetty thread instead of processing thread
```

Comment:
I want to limit the memory usage when scan segments, memory is the pain point of select query which buffers too many data in memory before return results to client.
So I decide to prepare the next batch when the client ask for it.

Another reason for using qtp instead of queryExecutor is that, the processing thread is valuable and usually number limited(because the direct memory is configed per processing thread). Usually the Scan query can last a very long time, it uses qtp thread anyway, so I think it is not a good idea to occupy one more processing thread for a very long time which may block other queries.